src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt

```diff
@@ -243,82 +219,192 @@ class IncomingPaymentHandler(val nodeParams: NodeParams, val db: IncomingPayment
                                 val failureMsg = IncorrectOrUnknownPaymentDetails(part.totalAmount, currentBlockHeight.toLong())
                                 when (part) {
                                     is HtlcPart -> actionForFailureMessage(failureMsg, part.htlc)
-                                    is PayToOpenPart -> actionForPayToOpenFailure(privateKey, failureMsg, part.payToOpenRequest) // NB: this will fail all parts, we could only return one
+                                    is WillAddHtlcPart -> actionForWillAddHtlcFailure(privateKey, failureMsg, part.htlc)
                                 }
                             }
                             pending.remove(paymentPart.paymentHash)
-                            return ProcessAddResult.Rejected(actions, incomingPayment)
+                            ProcessAddResult.Rejected(actions, incomingPayment)
                         }
-                        payment.amountReceived < payment.totalAmount -> {
+                        payment.amountReceived + payment.fundingFee < payment.totalAmount -> {
                             // Still waiting for more payments.
                             pending[paymentPart.paymentHash] = payment
-                            return ProcessAddResult.Pending(incomingPayment, payment)
+                            ProcessAddResult.Pending(incomingPayment, payment)
                         }
                         else -> {
-                            if (payment.parts.filterIsInstance<PayToOpenPart>().isNotEmpty()) {
-                                // We consider the total amount received (not only the pay-to-open parts) to evaluate whether or not to accept the payment
-                                val payToOpenFee = payment.parts.filterIsInstance<PayToOpenPart>().map { it.payToOpenRequest.payToOpenFeeSatoshis }.sum()
-                                nodeParams.liquidityPolicy.value.maybeReject(payment.amountReceived, payToOpenFee.toMilliSatoshi(), LiquidityEvents.Source.OffChainPayment, logger)?.let { rejected ->
-                                    logger.info { "rejecting pay-to-open: reason=${rejected.reason}" }
-                                    nodeParams._nodeEvents.emit(rejected)
-                                    val actions = payment.parts.map { part ->
-                                        val failureMsg = TemporaryNodeFailure
-                                        when (part) {
-                                            is HtlcPart -> actionForFailureMessage(failureMsg, part.htlc)
-                                            is PayToOpenPart -> actionForPayToOpenFailure(privateKey, failureMsg, part.payToOpenRequest) // NB: this will fail all parts, we could only return one
+                            val htlcParts = payment.parts.filterIsInstance<HtlcPart>()
+                            val willAddHtlcParts = payment.parts.filterIsInstance<WillAddHtlcPart>()
+                            when {
```

Comment:
I did not initially understand that, in the case where we are lacking inbound liquidity, we will enter this `when` twice:
- a first time in the `willAddHtlcParts.isNotEmpty()` branch, to *compute* the fee and ask for liquidity
- a second time in the `else` branch, to *verify* the fee and release the preimage

Reply:
Yes exactly, the first time triggers the open/splice, and then we will receive the `update_add_htlc` messages that couldn't initially be relayed (and were replaced by `will_add_htlc`) and go through the second step.